### PR TITLE
Eliminate ContinueWith from Renderer.cs. Fixes #6385

### LINKED ISSUE #6385: Blazor Index Out of Range Exception
### Describe the bug
Exception thrown: 'System.IndexOutOfRangeException' in System.Private.CoreLib.dll
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)

I have a periodic task that runs at 1Hz and components that get their data and update their state call StateHasChanged() at that frequency.

### To Reproduce
1. Using this version of ASP.NET Core: Blazor 0.7.0
2. Run this code:
```
            _updateTimer = new Timer((state) =>
            {
                try
                {
                    Updated();
                }
                catch(Exception ext)
                {
                    System.Diagnostics.Trace.WriteLine(ext.ToString());
                }
            }, null, TimeSpan.FromSeconds(1), TimeSpan.FromSeconds(1));
```
3. With these arguments:
Blazor Components that register for this event call StateHasChanged() when appropriate.
4. See error:
```
Exception thrown: 'System.IndexOutOfRangeException' in System.Private.CoreLib.dll
System.IndexOutOfRangeException: Index was outside the bounds of the array.
   at System.Collections.Generic.Dictionary`2.TryInsert(TKey key, TValue value, InsertionBehavior behavior)
   at Microsoft.AspNetCore.Blazor.Rendering.Renderer.AssignEventHandlerId(RenderTreeFrame& frame)
   at Microsoft.AspNetCore.Blazor.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForAttributeFrame(DiffContext& diffContext, Int32 oldFrameIndex, Int32 newFrameIndex)
   at Microsoft.AspNetCore.Blazor.RenderTree.RenderTreeDiffBuilder.AppendAttributeDiffEntriesForRange(DiffContext& diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
   at Microsoft.AspNetCore.Blazor.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForFramesWithSameSequence(DiffContext& diffContext, Int32 oldFrameIndex, Int32 newFrameIndex)
   at Microsoft.AspNetCore.Blazor.RenderTree.RenderTreeDiffBuilder.AppendDiffEntriesForRange(DiffContext& diffContext, Int32 oldStartIndex, Int32 oldEndIndexExcl, Int32 newStartIndex, Int32 newEndIndexExcl)
   at Microsoft.AspNetCore.Blazor.RenderTree.RenderTreeDiffBuilder.ComputeDiff(Renderer renderer, RenderBatchBuilder batchBuilder, Int32 componentId, ArrayRange`1 oldTree, ArrayRange`1 newTree)
   at Microsoft.AspNetCore.Blazor.Rendering.ComponentState.RenderIntoBatch(RenderBatchBuilder batchBuilder, RenderFragment renderFragment)
   at Microsoft.AspNetCore.Blazor.Rendering.Renderer.RenderInExistingBatch(RenderQueueEntry renderQueueEntry)
   at Microsoft.AspNetCore.Blazor.Rendering.Renderer.ProcessRenderQueue()
   at Microsoft.AspNetCore.Blazor.Rendering.Renderer.AddToRenderQueue(Int32 componentId, RenderFragment renderFragment)
   at Microsoft.AspNetCore.Blazor.Server.Circuits.CircuitSynchronizationContext.ExecuteSynchronously(TaskCompletionSource`1 completion, SendOrPostCallback d, Object state)
   at Microsoft.AspNetCore.Blazor.Server.Circuits.CircuitSynchronizationContext.Post(SendOrPostCallback d, Object state)
   at NG.Blazor.Client.Components.Settings.<OnAfterRender>b__10_0() in C:\compile\BlazorEcdis\NG.Blazor.Client\Components\Settings.cshtml:line 32
   at NG.Blazor.Client.DataService.<.ctor>b__7_0(Object state) in C:\compile\BlazorEcdis\NG.Blazor.Client\DataService.cs:line 31
```
### Expected behavior
Dictionary can be added to.

### Screenshots
N/A

### Additional context
Likely this could be addressed by changing the Dictionary that houses the _eventBindings (and _componentStateById?) to a Concurrent Dictionary here:
https://github.com/aspnet/AspNetCore/blob/436b5461ad0337aac90089aa7ccb61dd875808e4/src/Components/src/Microsoft.AspNetCore.Components/Rendering/Renderer.cs

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
This keeps the event handler disposal logic on the renderer's synchronization context, which should eliminate the hard-to-repro `IndexOutOfRangeException` reported in #6385.

As we've noted before, the logic in `Renderer.cs` is pretty intricate (methods rely on knowing details about the differing sync/async and error handling expectations of other methods, which you can only know about by reading the code extensively, hence all the comments). Hopefully we will improve on this in our planned refactoring.

### CODE REVIEW

File: src/Components/Components/src/Rendering/Renderer.cs [hunk 1]
```diff
@@ -622,8 +625,19 @@ private void RemoveEventHandlerIds(ArrayRange<int> eventHandlerIds, Task afterTa
...
+                    await afterTaskIgnoreErrors;
+                }
+                catch (Exception)
+                {
+                    // As per method contract, we're not error-handling the task.
+                    // That remains the caller's business.
+                }
+
+                // We know the next execution will complete synchronously, so nothing to await
+                _ = RemoveEventHandlerIdsAsync(eventHandlerIdsClone, Task.CompletedTask);
```
Comment (rynowak) [inline src/Components/Components/src/Rendering/Renderer.cs]: Is there a good reason to do this vs just extracting this L612-618 to a local function?

Comment (SteveSandersonMS) [reply src/Components/Components/src/Rendering/Renderer.cs]: This logic is how it already is prior to this PR. We could refactor further, but it’s not required for this PR.

**Review by rynowak — APPROVED** [review]

**Review by javiercn — COMMENTED** [review]
Looks good to me, does this change make the sync scenario more "allocaty"?

**Comment by SteveSandersonMS** [comment]
@javiercn You're right. This was a poor implementation - thanks for pointing it out. I've improved this in #9635.

**Pull request merged**